ui/src/app/components/new-user-satisfaction-survey-modal.spec.tsx

```diff
@@ -0,0 +1,129 @@
+import { shallow, ShallowWrapper } from 'enzyme';
+
+import {
+  CreateNewUserSatisfactionSurvey,
+  NewUserSatisfactionSurveySatisfaction,
+  SurveysApi,
+} from 'generated/fetch';
+
+import { ErrorMessage } from 'app/components/inputs';
+import {
+  registerApiClient,
+  surveysApi,
+} from 'app/services/swagger-fetch-clients';
+import { createNewUserSatisfactionSurveyStore } from 'app/utils/stores';
+
+import { SurveysApiStub } from 'testing/stubs/surveys-api-stub';
+
+import {
+  ADDITIONAL_INFO_MAX_CHARACTERS,
+  NewUserSatisfactionSurveyModal,
+} from './new-user-satisfaction-survey-modal';
+
+describe(NewUserSatisfactionSurveyModal.name, () => {
+  const setNewUserSatisfactionSurveyData = <
+    T extends keyof CreateNewUserSatisfactionSurvey
+  >(
+    attribute: T,
+    value: CreateNewUserSatisfactionSurvey[T]
+  ) => {
+    createNewUserSatisfactionSurveyStore.set({
+      newUserSatisfactionSurveyData: {
+        ...createNewUserSatisfactionSurveyStore.get()
+          .newUserSatisfactionSurveyData,
+        [attribute]: value,
+      },
+    });
+  };
+
+  const setValidNewUserSatisfactionSurveyData = () => {
+    setNewUserSatisfactionSurveyData(
+      'satisfaction',
+      NewUserSatisfactionSurveySatisfaction.VERYSATISFIED
+    );
+    setNewUserSatisfactionSurveyData('additionalInfo', '');
+  };
+
+  beforeEach(() => {
+    registerApiClient(SurveysApi, new SurveysApiStub());
+    setValidNewUserSatisfactionSurveyData();
+  });
+
+  const createShallowWrapper = ({
+    onCancel = () => {},
+    onSubmitSuccess = () => {},
+  } = {}) => {
+    return shallow(
+      <NewUserSatisfactionSurveyModal
+        onCancel={onCancel}
+        onSubmitSuccess={onSubmitSuccess}
+      />
+    );
+  };
+
+  const findSubmitButton = (wrapper: ShallowWrapper) => {
+    return wrapper.find({ children: 'Submit' });
+  };
+
+  it('should disable the submit button until a satisfaction value is selected', () => {
+    setNewUserSatisfactionSurveyData('satisfaction', undefined);
+    expect(
+      findSubmitButton(createShallowWrapper()).prop('disabled')
+    ).toBeTruthy();
+
+    setNewUserSatisfactionSurveyData(
+      'satisfaction',
+      NewUserSatisfactionSurveySatisfaction.VERYSATISFIED
+    );
+    expect(
+      findSubmitButton(createShallowWrapper()).prop('disabled')
+    ).toBeFalsy();
+  });
+
+  it('should disable the submit button if additional info exceeds the max length', () => {
+    setNewUserSatisfactionSurveyData(
+      'additionalInfo',
+      'A'.repeat(ADDITIONAL_INFO_MAX_CHARACTERS + 1)
+    );
+    expect(
+      findSubmitButton(createShallowWrapper()).prop('disabled')
+    ).toBeTruthy();
+
+    setNewUserSatisfactionSurveyData(
+      'additionalInfo',
+      'A'.repeat(ADDITIONAL_INFO_MAX_CHARACTERS)
+    );
+    expect(
+      findSubmitButton(createShallowWrapper()).prop('disabled')
+    ).toBeFalsy();
+  });
+
+  it('should disable the submit button while awaiting submission response', async () => {
```

Comment:
I'm assuming you mean the test and not the feature under test.

Javascript is single-threaded and runs a series of "event loops." Any time you `await` a promise, code following it will be run on a subsequent event, _even if the `await`ed promise immediately resolves_.

In this case, after running `findSubmitButton(wrapper).prop('onClick')();` we hit the code `await surveysApi().createNewUserSatisfactionSurvey(...);` which stops execution of the `onClick` function and returns to our test with a half-run `onClick`. When we `await` the returned promise (`async` functions always return a promise) in our test, the current event loop finishes and we move back to the `onClick` function. Once the `onClick` function finishes, our test code is added to the event loop because it was waiting for the `onClick` promise to finish before continuing and is immediately processed since there are no other pending events.

Happy to walk through this in person if you'd like!

Reply:
I'm not fully understanding here (for example: why was a promise created at all?) but our usual pattern for "flush the event loop" is to call `waitOneTickAndUpdate()` which makes it explicit what we're doing.  Is this equivalent, or is there more going on?

Reply:
> Why was a promise created at all

`onClick` is an `async` function, and all `async` functions return promises.

> our usual pattern for "flush the event loop" ...

We specifically do _not_ want to "flush the event loop" here since the window where the user is waiting for an API call only exists for one loop in a test environment.

I could be more explicit by manually stubbing `createNewUserSatisfactionSurvey` and `resolve`-ing its `Promise` response after we check the submit button is disabled (this would still require us to `await` the `onClick` response). That may make the test less brittle as well.

I can open a new PR to do that, if you'd like.

Reply:
Ok now I've got it.  Thanks for helping me through this. :)

**`prop('onClick')` resolves to `Button`'s `onClick()`!**  Whoa, I didn't know our test framework was sophisticated enough to do something like that.  That's the key piece that I was missing here.

So now the rest of it makes sense to me:
* onClick() fires and sets `submitting` immediately, but blocks at `await`
* therefore the Button is disabled, which we observe
* we resolve the promise, and onClick() can continue past `await`, setting `submitting` to false
* therefore the Button is enabled, which we observe

I don't know if it's necessary to resolve the promise explicitly, but I think I would have gotten here sooner with a comment or two, something like:

"onClick() sets `submitting` so the button is disabled until the API returns"
"the onClick promise has now resolved, so we are no longer submitting/disabled"

Reply:
Thanks, going forward I'll try adding clarifying comments when writing non-obvious async code 👍